### DISCUSSION TITLE
List with possible crash error codes

### DISCUSSION THREAD

**Discussion by knochenhans** [Quill-OS/quill#8] [Q&A] [answered]
Hi, I’m currently trying to run InkBox on my Touch 905B and I’m running into different crash error codes (2 and now 4) at boot. Is there a list available that list possible reasons for these kind of errors or what they mean?

**Comment by tux-linux** [comment]
No, but you can look at this script: https://github.com/Kobo-InkBox/kernel/blob/master/initrd/common/overlay-mount which calls the inkbox-splash script which prints the errors on the screen.
Errors 4 and 2 mean you've been getting a bunch of signature errors. Did it work before, or does it just happen like that after you've first flashed the image?

**Comment by knochenhans** [comment]
Thanks, that’s helpful! :) It’s the first time I flashed the file (http://23.163.0.39/bundles/inkbox/native/n905b/inkbox-1.7-n905b.xz) using dd onto a 4 GB card (the original 2 GB in my Kobo was too small). Maybe I’ll try redownloading the image, if this is signature related.
Could this also indicate possible compatibility problems between the sd card and the Kobo?

**Reply by tux-linux** [reply]
You seem to have flashed the correct file. What's its `sha256sum`? I have it here, we could verify. Otherwise, serial port output could be useful, too, although you maybe don't have the tools nor time to set it up.

**Reply by knochenhans** [reply]
I downloaded the file again and compared, both give me ``c6b419ede412d53647a61a51bcf63a6ed1b328e3414fa52f2713d7f0ecfa4db2``.
Thanks for the serial port tip, I might take a deeper look into that if it’s possible without having to solder anything.

**Reply by tux-linux** [reply]
Hmm yeah, that's strange. I have the same too.
Are you able to access the Diagnostics partition? Power on the device, then once the green light starts flashing, slide the power button a couple times for about 10 seconds.
Then, go into "Reset options" -> "Reset as root". It should take ~15 min. and then it should be good to go.
Otherwise, try reflashing the image and syncing your disks 10 or 20 times to be sure any writes cached have been written to the MMC.

**Reply by knochenhans** [reply]
Thanks, managed to get into the Diagnostics and currently waiting for the reset.

Before your last answer I got it working to a degree: From some point on, normal boot started to crash with error 6 but when I then inserted the USB cable (connected to my PC), the Kobo rebooted and booted successfully into the InkBox. So it basically seemed to work but I had to connect it via USB for a successful start every time. Let’s see what the reset brings.

**Reply by tux-linux** [reply]
Wow, that's strange. Well, at least you know one way to make it work. I don't know what happened ...
Keep me posted on the restore procedure, I'd like to know how it goes. Thanks.

**Reply by knochenhans** [reply]
Unfortunately, the device is now reproducibly crashing with error 5 at boot. It might be just my specific Kobo though which is borked, always had problems with it and could never update to anything above the original firmware version because of the new versions freezing, for example. Anyhow, I’ll try to get it back into that half working state above again and see how well it works :)

**Reply by tux-linux** [reply]
You've gotten a step closer! It now just can't succeed in verifying the rootfs squashfs archive.
Don't restore/reflash it yet.
Have you got a Linux computer?

**Reply by knochenhans** [reply]
Yes, I have access to a Linux PC :)

**Reply by tux-linux** [reply]
If you reset as root:
If you can, enter DFL (Direct Firmware Loader) mode by powering off the device, connecting it to a PC which will start it, and as soon as the green light starts flashing, slide both the power button and the "Home" button at the same time during ~5 secs. Then, a white-on-black splash should appear on the screen and say "DFL". The whole MMC should be exported over USB.
If it doesn't work, you can do the same procedure as non-root below.

If you reset as non-root:
Take out the MMC and put it on your computer (this is simpler, but can be annoying sometimes, that's why DFL might be better for this).

For both:
Mount the third partition of the storage device (e. g. `/dev/mmcblk0p3` or `/dev/sdb3`) on some mountpoint. Use this command to verify the rootfs archive digest against the rootfs with the public key that I'll paste below:
```
openssl dgst -sha256 -verify public.pem -signature /some/mount/point/overlaymount-rootfs.squashfs.dgst /some/mount/point/overlaymount-rootfs.squashfs
```
The public key, just paste it onto a file named `public.pem`:
```
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0RoRR11FdGnQnHNUGJy8
rNW9YGMCt5iqghAyeE8lUOtC4aQS9Uze/SUzjH8wQfsFd1U2Y83yv2IVzDOArYVl
6YmO4UySocw3CcuSATNaSL0sFm518hsMT6vw5RM32mjWAw3U5iymxvvbg+F7nzc1
sTNmEMueFAyoiA+0wAiDBncurBLpkYvBmKz9jrZFjOmI3KbEa9rdqiUQr//L2cHH
35QQvLBj4dx3IljFc+pi7ygVmLz06K9Rp/+Fy06H9sWO22Gtuq3vyChKxrRSO5AO
jK5i1xTwXtQpKWPpE/dCcDVBmsxX8VEcEJQtJe4FVe3XhZuhHNJ7FfFIVkv9+xG6
TQIDAQAB
-----END PUBLIC KEY-----
```
If everything goes well, it should output a `Verified OK` on the terminal. Otherwise, it will show `Verification Failure`. If that happens, then we know where the problem is. I'll guide you from there after.

BTW, I flashed the image yesterday to my Touch B and everything was fine with the stock image. I'd guess it could be a problem with your microSD card, maybe consider trying another one. Or it might be something else, a specific hardware quirk that isn't there on mine (although I'm not aware of any revisions on the Touch B itself).
**EDIT:** Error code 5 is about the utility root filesystem, actually. Sorry for the mistake. I edited the commands below, they're pretty similar.

**Reply by knochenhans** [reply]
Thanks, running the command after mounting the rootfs partition gives me a ``Verification Failure``.

If it’s any help, here are the sha256sum for the two squashfs files in partition’s root folder:
- overlaymount-rootfs.squashfs: ``88a8305aaa52d206eb5c866fe37838d45c76c21633022fead2c863137a867739``
- rootfs.squashfs: ``f0b16c7738cc93e0a8bbcd3cae14a24b3ba5bf2855fe691a1db2a4da839b9abc``

**Reply by tux-linux** [reply]
Everything seems normal, I've got the same from my side.
What's the sha256sum for `overlaymount-rootfs.squashfs.dgst`?
Mine:
```
kobo:/mnt# sha256sum overlaymount-rootfs.squashfs.dgst
b0c8cd7170b9af3df7088ea42f3640313ec3f3e59dd6eab2ecce33e68d93ee06  overlaymount-rootfs.squashfs.dgst
```
And what was the `openssl` command you ran? It can get pretty confusing sometimes.

**Reply by knochenhans** [reply]
Hmm, mine looks like this:
```
sha256sum overlaymount-rootfs.squashfs.dgst 
b1a2584f440a9aa31e90903d2c4da7663a35dc516ba60461b9dbeafc27a767e8  overlaymount-rootfs.squashfs.dgst
```

My ``openssl`` command from above looked like this:
```
openssl dgst -sha256 -verify /tmp/public.pem -signature /run/media/andre/rootfs/overlaymount-rootfs.squashfs.dgst /run/media/andre/rootfs/overlaymount-rootfs.squashfs
```

(the `rootfs.squashfs` verified fine by the way)

**Reply by tux-linux** [reply]
Replace `overlaymount-rootfs.squashfs.dgst` with the joined zipped one, then try verifying it again.
[dgst.zip](https://github.com/Kobo-InkBox/inkbox/files/7780937/dgst.zip)

**Reply by knochenhans** [reply]
Thanks, that worked. I just went ahead after that and tried to restart the Kobo: This time, InkBox booted without problems. Strangely enough, it crashed with Error code 6 on the next reboot. Another reboot later it boots fine though.

**Reply by tux-linux** [reply]
The only time where it should display a code 6 error would be when you have encrypted storage setup and you type 5 times a wrong passphrase. The device then "locks" itself for a day and won't boot until the day has passed, showing error code 6 every time until the limit is reached.
See here: https://github.com/Kobo-InkBox/kernel/blob/199f74aaa60554136f48da239f5fd39ece925524/initrd/common/rcS-root#L112-L122
But you obviously didn't set up any of that, AFAIK, and this is a complete mystery ;)

**Reply by knochenhans** [reply]
Just a short note: I’m giving up for now. After the short success yesterday I haven’t been able to successfully boot the system, about 50% of boots crash with code 6, the other 50% crash earlier in the process with error 2 or 3. I guess I’ll try again with another SD card as soon as I get my hands on one. Many thanks for your help for now :)

**Comment by tux-linux** [comment, marked as answer]
Now there is: https://inkbox.ddns.net/wiki/index.php?title=Boot_error_codes
